ts/client/storageEngine/transactions/db.ts

```diff
@@ -0,0 +1,197 @@
+            transaction: txn
+        };
+
+        await this.toDB(txnMessage, txnStatus);
+        return txnStatus;
+    }
+
+    public async submitted(
+        msgOrTxn: TransationMessageOrObject,
+        { batchID, l1TxnHash, l1BlockIncluded }: SubmitMeta
+    ): Promise<TransactionStatus> {
+        const msg = this.getMessage(msgOrTxn);
+        const txnStatus = await this.getStatusOrFail(msg);
+        this.validateStatusTransition(txnStatus.status, Status.Submitted);
+        const submittedTxnStatus = {
+            ...txnStatus,
+            status: Status.Submitted,
+            batchID,
+            l1TxnHash,
+            l1BlockIncluded
+        };
+        await this.toDB(msg, submittedTxnStatus);
+        return submittedTxnStatus;
+    }
+
+    public async finalized(
+        msgOrTxn: TransationMessageOrObject
+    ): Promise<TransactionStatus> {
+        return this.transition(this.getMessage(msgOrTxn), Status.Finalized);
+    }
+
+    public async failed(
+        msgOrTxn: TransationMessageOrObject,
+        detail: string
+    ): Promise<TransactionStatus> {
+        return this.transition(
+            this.getMessage(msgOrTxn),
+            Status.Failed,
+            detail
+        );
+    }
+
+    public async sync(
+        txn: OffchainTx,
+        { batchID, l1TxnHash, l1BlockIncluded, finalized }: SyncMeta
+    ): Promise<TransactionStatus> {
+        const txnMessage = txn.message();
+        await this.throwIfAlreadyExists(txnMessage);
+
+        const status = finalized ? Status.Finalized : Status.Submitted;
+        const txnStatus = {
+            transaction: txn,
+            status,
+            batchID,
+            l1TxnHash,
+            l1BlockIncluded
+        };
+        await this.toDB(txnMessage, txnStatus);
+        return txnStatus;
+    }
+
+    public async count(): Promise<number> {
+        const stream = this.db.createKeyStream();
+        let count = 0;
+        for await (const _ of stream) {
+            count++;
+        }
+        return count;
+    }
+
+    private async throwIfAlreadyExists(txnMessage: string) {
+        try {
+            const itemFound = await this.get(txnMessage);
+            if (itemFound) {
+                throw new TransactionAlreadyExists(txnMessage);
+            }
+        } catch (error) {
+            if (error.name !== "NotFoundError") {
+                throw error;
+            }
+        }
+    }
+
+    private getMessage(msgOrTxn: TransationMessageOrObject): string {
+        if (typeof msgOrTxn === "string") {
+            return msgOrTxn;
+        }
+        return msgOrTxn.message();
+    }
+
+    private async getStatusOrFail(message: string): Promise<TransactionStatus> {
+        try {
+            return await this.get(message);
+        } catch (error) {
+            if (error.name === "NotFoundError") {
+                throw new TransactionDoesNotExist(message);
+            } else {
+                throw error;
+            }
```

Comment:
#nitpick `else` block is not needed here. Same as:
```typescript
            if (error.name === "NotFoundError") {
                throw new TransactionDoesNotExist(message);
            }
            throw error;
```
I generally prefer the above form to reduce nesting but it is a stylistic choice. Good as is.